src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -355,36 +367,37 @@ public <S extends DEREncodable> S decode(String str, Class<S> tClass) {
     }
 
     /**
-     * Decodes and returns the specified class for the given
-     * {@link InputStream}.  The class must extend {@link DEREncodable} and be
-     * an appropriate class for the PEM type.
+     * Decodes and returns a {@code DEREncodable} of the specified class for the
+     * given {@link InputStream}. {@code tClass} must be an appropriate class
+     * for the PEM type.
      *
      * <p> This method reads from the {@code InputStream} until the end of
-     * the PEM footer or the end of the stream. If an I/O error occurs,
+     * a PEM footer or the end of the stream. If an I/O error occurs,
      * the read position in the stream may become inconsistent.
      * It is recommended to perform no further decoding operations
      * on the {@code InputStream}.
      *
-     * <p> If the class parameter is {@code PEMRecord.class},
-     * a {@linkplain PEMRecord} is returned containing the
-     * type identifier and Base64 encoding. Any non-PEM data preceding
-     * the PEM header will be stored in {@code leadingData}.  Other
-     * class parameters will not return preceding non-PEM data.
+     * <p> If the class parameter is {@code PEM.class}, a {@link PEM} object is
+     * returned containing the type identifier, Base64-encoded data, and any
+     * leading data preceding the PEM header. For {@code DEREncodable} types
+     * other than {@code PEM}, leading data is ignored and not returned as part
+     * of the DEREncodable object.
      *
      * <p> If no PEM data is found, an {@code IllegalArgumentException} is
      * thrown.
      *
      * @param <S> Class type parameter that extends {@code DEREncodable}.
      * @param is an InputStream containing PEM data
-     * @param tClass the returned object class that implements
+     * @param tClass the returned object class that extends or implements
      *   {@code DEREncodable}.
      * @return a {@code DEREncodable} typecast to {@code tClass}
      * @throws IOException on IO or PEM syntax error where the
-     * {@code InputStream} did not complete decoding.
-     * @throws EOFException at the end of the {@code InputStream}
+     *   {@code InputStream} did not complete decoding.
+     * @throws EOFException no PEM data found or unexpectedly reached the
+     *   end of the {@code InputStream}
      * @throws IllegalArgumentException on error in decoding
-     * @throws ClassCastException if {@code tClass} is invalid for the PEM type
-     * @throws NullPointerException when any input values are null
+     * @throws ClassCastException if {@code tClass} does not represent the PEM type
```

Comment:
It's a little odd this throws a `ClassCastException`. This seems more like an `IllegalArgumentException` to me because you are passing in the wrong type.

Reply:
@wangweij pushed for CCE instead of IAE.  It is a valid argument, but cannot be cast to that object.  I think one can make a case for both exception types, but CCE was a bit stronger.